### PR TITLE
sort: don't rely on serde-json for extsort

### PR DESCRIPTION
**Pull request by miDeb**
It is much faster to just write the lines to disk, separated by \n
(or \0 if zero-terminated is enabled), instead of serializing to json.

external_sort now knows of the Line struct instead of interacting with
it using the ExternallySortable trait. Similarly, it now uses the
crash_if_err! macro to handle errors, instead of bubbling them up.

Some functions were changed from taking &[Line] as the input to taking
an Iterator<Item = Line>. This removes the need to collect to a Vec
when not necessary.

@electricboogie I think you're most familiar with the code here, would you mind reviewing this? Sorry that this is such a huge change.

### CODE REVIEW

**Comment by sylvestre** [comment]
Do you have some perf wins?

**Comment by miDeb** [comment]
Yes, this should be about 80% faster, but it is hard to say because we previously used more memory than what the flag told us to. I updated BENCHMARKING.md with benchmarking instructions.

We're still multiple times slower than GNU sort, especially on large files.

**Comment by sylvestre** [comment]
in benchmark, could you please provide some hyperfine examples to reproduce the improvement? thanks

File: src/uu/sort/BENCHMARKING.md [hunk 1]
```diff
@@ -69,6 +69,13 @@ Run `cargo build --release` before benchmarking after you make a change!
 
 -   Benchmark numeric sorting with hyperfine: `hyperfine "target/release/coreutils sort shuffled_numbers_si.txt -h -o output.txt"`.
 
+## External sorting
+
+Try running commands with the `-S` option set to an amount of memory to be used, such as `1M`. Additionally, you could try sorting
+huge files (ideally multiple Gigabytes) with `-S`. Creating such a large file can be achieved by running `cat shuffled_wordlist.txt | sort -R >> shuffled_wordlist.txt`
+multiple times (this will add the contents of `shuffled_wordlist.txt` to itself).
+Example: Run `hyperfine "target/release/coreutils sort shuffled_wordlist.txt -S 1M"`
```
Comment (sylvestre) [inline src/uu/sort/BENCHMARKING.md:77]:
```suggestion
Example: Run `hyperfine './target/release/coreutils sort shuffled_wordlist.txt -S 1M' 'sort shuffled_wordlist.txt -S 1M'`
`
```

**Comment by electricboogie** [comment]
Conceptually, this works, but I think you won't get the benefit of any cache as you do a final compare_by as you iterate(call next()) through the returned iter (the final pass).  So, it may perform worse on certain inputs, but I doubt it.

However, I think we may be prematurely optimizing here.  It may be the case that we find that we should write the entire struct to disk as, say, it is fastest for much larger inputs, or we may find another use case in the Line struct that is worth the bytes.  

My take is -- I think the real problem is the memory usage.  We are still consuming 3 to 4x more memory than GNU because of the size of the Line struct.  Maybe we get our data structures/the Line struct to a place we are happy with first?  My opinion is that you were right the first time -- we need to push out the NumCache and Field/Other stuff to a separate structure.  I suggested a few possible (naive!) solutions on the issue page.  I'd be pleased to assist anywhere you'd need me.

**Comment by miDeb** [comment]
I'm reasonably sure that reconstructing the cache at the last pass is much faster than serializing it to json. We also need to reduce memory usage, that's for sure. I promise I'll look into that :) let's see when I have time for this though 😅

I'll have to find out how to reliably measure memory usage, because at some point it looked like we're using less memory than GNU, which doesn't make much sense to me.

There's one bug that I'm fixing here that made the current version unusable though (not related to the json stuff), which is that we were always reading all lines before sorting, making all extsort useless (everything was loaded into memory anyways). You can't sort a big (something like 2GB) file with the current implementation. I also made the estimation on how much memory we're using more correct.

We could also leverage the current implementation for -m for the final merge step (to reduce code duplication, and what we have there (a `BinaryHeap`) should be faster than the algorithm in extsort when considering the algorithmic complexity)

**Comment by electricboogie** [comment]
Totally agree extsort is not yet ready for prime time -- memory usage looks like a V shape in Valgrind.  I was going to do some work to module-ize the extsort::sort_by implementation to accept chunks.  But after screwing around with that for awhile, I thought 2 things should probably come first: 1) we break input into chunks when feeding from the buffer, and 2) we make a sure we are happy with our data structures, because who wants to rework the extsort again.

File: src/uu/sort/src/external_sort/mod.rs [hunk 2]
```diff
@@ -91,205 +62,150 @@ where
...
+    // We will have to have the entire iter in memory sometime right?
+    // Set minimum to the size of the writer buffer, ~8K
+
+    const MINIMUM_READBACK_BUFFER: usize = 8200;
+    let right_sized_buffer = settings
+        .buffer_size
+        .checked_div(iter.chunks)
+        .unwrap_or(settings.buffer_size);
+    iter.max_per_chunk = if right_sized_buffer > MINIMUM_READBACK_BUFFER {
+        right_sized_buffer
```
Comment (sylvestre) [inline src/uu/sort/src/external_sort/mod.rs:152]: would it be possible to add a test for this case?

File: src/uu/sort/src/external_sort/mod.rs [hunk 3]
```diff
@@ -91,205 +62,150 @@ where
...
-{
+    max_bytes: usize,
+    settings: &GlobalSettings,
+) -> usize {
     let mut total_read = 0;
     let mut bytes_read = 0;
-    for line in BufReader::new(file).lines() {
-        let line_s = line?;
+    for line in BufReader::new(file).split(if settings.zero_terminated {
+        b'\0'
```
Comment (sylvestre) [inline src/uu/sort/src/external_sort/mod.rs:196]: ditto

File: src/uu/sort/src/external_sort/mod.rs (hunk 2 above)
Comment (miDeb) [reply src/uu/sort/src/external_sort/mod.rs:152]: done...

File: src/uu/sort/src/external_sort/mod.rs (hunk 3 above)
Comment (miDeb) [reply src/uu/sort/src/external_sort/mod.rs:196]: ...and done.

**Pull request merged**